### PR TITLE
Added flux demo

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 1]
```diff
@@ -705,7 +705,11 @@ def run(
         self._create_timing_cache(
             builder_config, self.compilation_settings.timing_cache_path
         )
-
+        # TODO: Memory control prototyping. Under discussion
+        self.module.to("cpu")
+        torch.cuda.empty_cache()
+        del self.module
+        gc.collect()
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:712]: you can use delete module call here

File: demo/flux_demo.py
Comment (narendasan) [inline demo/flux_demo.py:1]: move to examples

**Comment by narendasan** [comment]
Can the app display the inference time, might be nice to have some stats rendered live as you generate

File: py/torch_tensorrt/runtime/_cudagraphs.py [hunk 2]
```diff
@@ -69,52 +69,58 @@ def __init__(self, compiled_module: torch.nn.Module) -> None:
...
+                "The model contains submodules that require a dynamic output allocator at runtime, which is incompatible with CUDA Graphs. Please disable CUDA Graphs."
+            )
+        if "_run_on_acc" in name:
+            num_trt_module += 1
+        elif "_run_on_gpu" in name:
+            num_torch_module += 1
+
+    if num_torch_module > 0:
+        # Set whole cudagraphs mode and returns wrapped module
+        _PY_RT_CUDAGRAPHS = CudaGraphsMode.WHOLE_GRAPH_CUDAGRAPHS
```
Comment (narendasan) [inline py/torch_tensorrt/runtime/_cudagraphs.py:103]: This messes with global state in a weird way since this will toggle the entire runtime's mode.

File: examples/apps/flux-demo.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    # "reuse_cached_engines": True,
+    # "timing_cache_path": "/home/engine_cache/flux.bin",
+    # "engine_cache_size": 40 * 1 << 30,
+    # "enable_weight_streaming": True,
+    # "weight_streaming_budget": 8 * 1 << 30
+    # "enable_cuda_graph": True,
+}
+
+trt_gm = torch_tensorrt.MutableTorchTensorRTModule(backbone, **settings)
+trt_gm.set_expected_dynamic_shape_range((), dynamic_shapes)
```
Comment (narendasan) [inline examples/apps/flux-demo.py:56]: why is the empty tuple here?

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 4]
```diff
@@ -252,7 +209,7 @@ def forward(a, b, c=0, d=0):
 
         self.refit_state.set_state(RefitFlag.NEEDS_RECOMPILE)
 
-    def _get_total_dynamic_shapes(self) -> Union[dict[str, Any], None]:
+    def _get_total_dynamic_shapes(self) -> dict[str, Any] | None:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:212]: Use the union syntax until 3.9 support is dropped

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 5]
```diff
@@ -491,14 +479,23 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
...
         result = self.gm(*args, **kwargs)
         # Storing inputs and outputs for verification when the state is unknown
         self.run_info = (args, kwargs, result)
         return result
 
-    def to(self, device: str) -> None:
-        logger.warning("Original PyTorch model is moved. CPU offload may failed.")
-        self.original_model.to(device)
+    def to(self, *args: Any, **kwargs: Any) -> None:
+        logger.warning(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:491]: Probably should tell users that the module was not actually moved

File: py/torch_tensorrt/runtime/_cudagraphs.py [hunk 6]
```diff
@@ -68,51 +68,66 @@ def __init__(self, compiled_module: torch.nn.Module) -> None:
...
-                    "Please consider dynamo if there is graph breaks. Using runtime cudagraphs mode"
-                )
-            # Enable cudagraphs for TRT submodule
-            set_cudagraphs_mode(True)
+        if isinstance(self.compiled_module, torch_tensorrt.MutableTorchTensorRTModule):
+            self.old_module = self.compiled_module.gm
+            self.compiled_module.gm = get_cuda_graph_module(self.compiled_module.gm)
             return self.compiled_module
 
+        return get_cuda_graph_module(self.compiled_module)
```
Comment (narendasan) [inline py/torch_tensorrt/runtime/_cudagraphs.py:80]: add the else:

**Review by narendasan — COMMENTED** [review]
Lets separate all the CPU offload API stuff into another API

**Comment by narendasan** [comment]
Also rebase

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 7]
```diff
@@ -550,15 +551,6 @@ def compile(
                 "`immutable_weights` must be False when `refit_identical_engine_weights` is True."
             )
 
-    if (
```
[the remark this answers is not in the document]
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/_compiler.py:553]: Yes

File: examples/apps/flux-demo.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    pipe.fuse_lora()
+    pipe.unload_lora_weights()
+    print("LoRA loaded! Begin refitting")
+    generate_image(["Test"], 2)
+    print("Refitting Finished!")
+
+
+generate_image(["Test"], 2)
+# load_lora("")
+# generate_image(["A golden retriever holding a sign to code"], 2)
```
Comment (peri044) [inline examples/apps/flux-demo.py:101]: remove comment

File: py/torch_tensorrt/dynamo/_refit.py [hunk 9]
```diff
@@ -462,25 +496,39 @@ def refit_module_weights(
...
 
+        del engine
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    # TODO: Memory control prototyping. Under discussion
+    if settings.offload_module_to_cpu:
+        del new_partitioned_module
+        gc.collect()
+        torch.cuda.empty_cache()
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_refit.py:530]: use delete_module call here

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 10]
```diff
@@ -624,18 +621,58 @@ def _check_tensor_shapes_with_dynamic_shapes(
...
+                for v in obj:
+                    resursivly_serialize_dynamic_shape(v)
+
+        resursivly_serialize_dynamic_shape(self.arg_dynamic_shapes)
+        resursivly_serialize_dynamic_shape(self.kwarg_dynamic_shapes)
+
+    def deserialize_dynamic_shapes(self) -> None:
+        dims = self.serializable_dynamic_shapes_dims
+
+        def resursivly_deserialize_dynamic_shape(obj: Any) -> None:
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:647]: recursively*

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 11]
```diff
@@ -345,9 +302,28 @@ def refit_gm(self) -> None:
...
+                allow_complex_guards_as_runtime_asserts=self.allow_complex_guards_as_runtime_asserts,
+            )
+        else:
+            return torch.export.export(
+                self.original_model,
+                self.arg_inputs,
+                kwargs=self.kwarg_inputs,
+                dynamic_shapes=self._get_total_dynamic_shapes(),
+                strict=self.strict,
+            )
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:325]: I think let's do a try catch style here where we try torch.export.export first and then _trace.export. I found _export doesn't work with modelopt recently.

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 12]
```diff
@@ -345,9 +302,28 @@ def refit_gm(self) -> None:
             in_place=True,
         )
 
-        self.original_model.cpu()
+        self.original_model.to("cpu")
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:305]: use the CPU_DEVICE variable in the utils

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py (hunk 11 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:325]: Catch style will make the process slower because it needs to try one and the other, right? My thoughts here is have the user decide whether they have to use allow_complex_guards_as_runtime_asserts, which only _export has. If that is absolute necessary, then we cannot use torch.export.export.

File: examples/apps/flux-demo.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
+import argparse
+import re
+import time
+
+import gradio as gr
+import modelopt.torch.quantization as mtq
+import register_sdpa
```
Comment (peri044) [inline examples/apps/flux-demo.py]: We can avoid copying the `register_sdpa.py` and `sdpa_converter.py` files by doing this 
```py
import sys
import os

# Register SDPA as a standalone operator. Converter and lowering pass are defined in register_sdpa.py
sys.path.append(os.path.join(os.path.dirname(__file__), '../dynamo'))
from register_sdpa import *
```

File: examples/apps/NGRVNG.safetensors
Comment (peri044) [inline examples/apps/NGRVNG.safetensors:1]: what does this file do ?

File: examples/dynamo/torch_export_flux_dev.py [hunk 14]
```diff
@@ -112,6 +112,8 @@
     min_block_size=1,
     use_fp32_acc=True,
     use_explicit_typing=True,
+    use_python_runtime=True,
```
Comment (peri044) [inline examples/dynamo/torch_export_flux_dev.py]: can we default to using C++ runtime (use_python_runtime=False) ?

File: examples/dynamo/torch_export_flux_dev.py [hunk 15]
```diff
@@ -120,13 +122,13 @@
...
-backbone.to("cpu")
 pipe.to(DEVICE)
-torch.cuda.empty_cache()
+backbone.to("cpu")
 pipe.transformer = trt_gm
+del ep
+torch.cuda.empty_cache()
 pipe.transformer.config = config
-
+trt_gm.device = torch.device("cuda")
```
Comment (peri044) [inline examples/dynamo/torch_export_flux_dev.py]: Can we use the offload_module_to_cpu=True to handle this block of code ?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 16]
```diff
@@ -912,7 +913,7 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
         parse_graph_io(submodule, subgraph_data)
         dryrun_tracker.tensorrt_graph_count += 1
         dryrun_tracker.per_subgraph_data.append(subgraph_data)
-
+        torch.cuda.empty_cache()
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py:918]: is this needed here ?

File: py/torch_tensorrt/dynamo/_refit.py [hunk 17]
```diff
@@ -341,7 +370,7 @@ def refit_module_weights(
 
     # Iterate over all components that can be accelerated
     # Generate the corresponding TRT Module for those
-
+    new_weight_module.module().to(CPU_DEVICE)
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_refit.py:373]: Is the new_weight_module the updated weights module that user provides ?

File: py/torch_tensorrt/dynamo/_refit.py [hunk 18]
```diff
@@ -444,33 +473,41 @@ def refit_module_weights(
...
             setattr(compiled_module, f"{name}_engine", refitted_engine)
 
+        del engine
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    deallocate_module(new_partitioned_module)
+
     if verify_output and arg_inputs is not None:
+        new_gm.to(torch.cuda.current_device())
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_refit.py]: We should ensure we use the device that's passed in via args/Compilation settings or default device and not call rely on torch cuda calls unless it is needed.

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 19]
```diff
@@ -356,25 +351,37 @@ def compile(self) -> None:
...
             arg_inputs=self.arg_inputs,
             kwarg_inputs=self.kwarg_inputs,
-            **self.settings.__dict__,
+            immutable_weights=False,
+            use_python_runtime=self.use_python_runtime,
+            **self.additional_settings,
         )
-        self.original_model.cpu()
+        self.original_model.to("cpu")
         torch.cuda.empty_cache()
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py]: use deallocate module

File: tools/perf/Flux/register_sdpa.py [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+from torch_tensorrt.dynamo._settings import CompilationSettings
+from torch_tensorrt.dynamo.conversion.aten_ops_converters import args_bounds_check
+from torch_tensorrt.dynamo.lowering import TORCH_TRT_DECOMPOSITIONS
+from torch_tensorrt.dynamo.lowering.passes._aten_lowering_pass import (
+    _aten_lowering_pass,
+)
+from torch_tensorrt.dynamo.lowering.passes.pass_utils import (
+    clean_up_graph_after_modifications,
+)
+
```
Comment (peri044) [inline tools/perf/Flux/register_sdpa.py]: use them from examples

File: tools/perf/Flux/flux_quantization.py
Comment (peri044) [inline tools/perf/Flux/flux_quantization.py:1]: I think we should avoid copying the whole model scripts for measuring perf. Try using the sys.path approach and importing the model and just a perf loop. something like 
```py
import sys
import os
sys.path.append(torchtrt_root + "examples/dynamo/apps")
from flux_demo import *
model = <insert FLUX model (fp16 or fp8) > 
results = measure_flux_perf(.... )
```

File: examples/apps/NGRVNG.safetensors
Comment (narendasan) [reply examples/apps/NGRVNG.safetensors:1]: Remove this file. We can check in other people weights into our repo

File: examples/apps/flux-demo.py [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+            model_dropdown = gr.Dropdown(
+                choices=["Torch Model", "Torch-TensorRT Accelerated Model"],
+                value="Torch-TensorRT Accelerated Model",
+                label="Model Variant",
+            )
+
+            lora_upload_path = gr.Textbox(
+                label="LoRA Path",
+                placeholder="Enter the LoRA checkpoint path here",
+                value="/home/TensorRT/examples/apps/NGRVNG.safetensors",
```
Comment (narendasan) [inline examples/apps/flux-demo.py]: Find a way to download this directly from HF. The APIs should be able to support it

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 16 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/_compiler.py:918]: Yes, there are some intermediate cache that could be released before the interpreter

File: py/torch_tensorrt/dynamo/_refit.py (hunk 17 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/_refit.py:373]: yes

File: examples/apps/flux_demo.py [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+
+    pipe = FluxPipeline.from_pretrained(
+        "black-forest-labs/FLUX.1-dev",
+        torch_dtype=torch.float16,
+    ).to(torch.float16)
+
+    if args.debug:
+        pipe.transformer = FluxTransformer2DModel(
+            num_layers=1, num_single_layers=1, guidance_embeds=True
+        ).to(torch.float16)
```
Comment (peri044) [inline examples/apps/flux_demo.py]: remove this
